File: server_app/pdf/settlement_notes.py
from html import escape


def settlement_note_markup(statement, columns, lines=()):
    by_species = {}
    for column in columns:
        by_species.setdefault(column["species"], []).append(column)
    rate_entries = []
    group_count = len(by_species)
    for index, (species, entries) in enumerate(by_species.items(), start=1):
        descriptors = list(dict.fromkeys(rate_descriptor(entry) for entry in entries))
        prefix = f"{index}）" if group_count > 1 else ""
        rate_entries.append(f"{prefix}{species} {'；'.join(descriptors)}")
    rate_markup = note_entry_markup("收费标准：", "、".join(rate_entries)) if rate_entries else ""
    custom_markup = custom_billing_note_markup(lines)
    expiry_note = str(statement.get("notes") or "").strip()
    expiry_markup = note_entry_markup("伦理到期提示：", expiry_note) if expiry_note else ""
    return '<div class="note-line">' + rate_markup + custom_markup + expiry_markup + "</div>"


def custom_billing_note_markup(lines):
    details = {}
    for line in lines or ():
        for item in line.get("iacucBreakdown") or ():
            if not item.get("customBilling"):
                continue
            billing_unit = str(item.get("billingUnit") or "")
            key = (
                str(item.get("iacuc") or "").strip().upper(),
                str(item.get("customBillingSegmentId") or "").strip(),
                str(item.get("customBillingStartDate") or "").strip(),
                str(item.get("customBillingEndDate") or "").strip(),
                as_number(item.get("unitPrice")),
                billing_unit,
                str(item.get("customBillingNote") or "").strip(),
            )
            current = details.setdefault(
                key,
                {
                    "iacuc": key[0],
                    "startDate": key[2],
                    "endDate": key[3],
                    "quantity": as_number(
                        item.get("animalCount") if billing_unit == "animal_day" else item.get("cageCount")
                    ),
                    "unitPrice": key[4],
                    "billingUnit": billing_unit,
                    "note": key[6],
                    "amount": 0,
                },
            )
            current["amount"] += as_number(item.get("payableAmount", item.get("amount")))
    if not details:
        return ""
    rows = []
    for item in sorted(details.values(), key=lambda value: (value["iacuc"], value["startDate"], value["endDate"])):
        unit = "只" if item["billingUnit"] == "animal_day" else "笼"
        period = f"{item['startDate'] or '-'} 至 {item['endDate'] or '-'}"
        note = f"，{item['note']}" if item["note"] else ""
        rows.append(
            f"{item['iacuc']}：{period}，每日 {number_text(item['quantity'])}{unit}，"
            f"{number_text(item['unitPrice'])}元/{unit}/日，本期 {number_text(item['amount'])}元{note}"
        )
    return "".join(note_entry_markup("自定义收费：" if index == 0 else "", row) for index, row in enumerate(rows))


def note_entry_markup(title, detail):
    return (
        '<div class="note-entry">'
        + (f"<strong>{html_text(title)}</strong>" if title else "")
        + f'<span class="note-detail">{html_text(detail)}</span></div>'
    )


def rate_descriptor(column):
    unit = "只" if column["unit"] == "animal_day" else "笼"
    if (
        column["species"] == "小鼠"
        and column["tiered"]
        and column["hasTieredCharge"]
        and column["overageUnitPrice"] > 0
        and column["overageUnitPrice"] != column["unitPrice"]
    ):
        return f"笼位数≤160，{number_text(column['unitPrice'])}元/笼/日；笼位数＞160，{number_text(column['overageUnitPrice'])}元/笼/日"
    return f"{number_text(column['unitPrice'])}元/{unit}/日"


def as_number(value):
    try:
        return float(value or 0)
    except (TypeError, ValueError):
        return 0.0


def number_text(value):
    number = as_number(value)
    return str(int(number)) if number.is_integer() else str(round(number, 2))


def html_text(value):
    return escape(str(value or ""))

File: src/react/print/settlementNotes.ts
import type { BillingStatementLine } from "../api/contracts";
import { escapeHtml, money, normalizeIacuc, numberText, speciesLabelFor } from "./settlementSupport";

type SettlementNoteColumn = {
  speciesLabel: string;
  billingUnit: string;
  unitPrice: number;
  overageUnitPrice: number;
  tiered: boolean;
  hasTieredCharge: boolean;
};

type CustomBillingBreakdown = {
  iacuc?: string;
  species?: string;
  billingItem?: string;
  animalCount?: number;
  cageCount?: number;
  billingUnit?: string;
  unitPrice?: number;
  amount?: number;
  payableAmount?: number;
  customBilling?: boolean;
  customBillingSegmentId?: string;
  customBillingStartDate?: string;
  customBillingEndDate?: string;
  customBillingNote?: string;
};

export function settlementNotesMarkup(
  columns: SettlementNoteColumn[],
  notes: string | undefined,
  lines: BillingStatementLine[] = [],
) {
  const ratesBySpecies = new Map<string, SettlementNoteColumn[]>();
  columns.forEach((column) => {
    const entries = ratesBySpecies.get(column.speciesLabel) || [];
    entries.push(column);
    ratesBySpecies.set(column.speciesLabel, entries);
  });
  const rateEntries = [...ratesBySpecies.entries()].map(([species, entries], index, groups) => {
    const descriptors = [...new Set(entries.map(rateDescriptor))];
    const prefix = groups.length > 1 ? `${index + 1}）` : "";
    return `${prefix}${species} ${descriptors.join("；")}`;
  });
  const rateMarkup = rateEntries.length ? noteEntryMarkup("收费标准：", rateEntries.join("、")) : "";
  const customMarkup = customBillingNoteMarkup(lines);
  const expiryNote = String(notes || "").trim();
  const expiryMarkup = expiryNote ? noteEntryMarkup("伦理到期提示：", expiryNote) : "";
  return `<div class="note-line">${rateMarkup}${customMarkup}${expiryMarkup}</div>`;
}

function customBillingNoteMarkup(lines: BillingStatementLine[]) {
  const details = new Map<
    string,
    {
      iacuc: string;
      startDate: string;
      endDate: string;
      quantity: number;
      unitPrice: number;
      billingUnit: string;
      species: string;
      note: string;
      amount: number;
    }
  >();
  lines.forEach((line) =>
    (line.iacucBreakdown || []).forEach((raw) => {
      const item = raw as CustomBillingBreakdown;
      if (!item.customBilling) return;
      const iacuc = normalizeIacuc(item.iacuc);
      const startDate = String(item.customBillingStartDate || "");
      const endDate = String(item.customBillingEndDate || "");
      const unitPrice = Number(item.unitPrice || 0);
      const billingUnit = String(item.billingUnit || "cage_day");
      const species = speciesLabelFor(item);
      const note = String(item.customBillingNote || "");
      const key = [iacuc, item.customBillingSegmentId || "", startDate, endDate, unitPrice, billingUnit, note].join(
        "|",
      );
      const current = details.get(key) || {
        iacuc,
        startDate,
        endDate,
        quantity: Number(billingUnit === "animal_day" ? item.animalCount || 0 : item.cageCount || 0),
        unitPrice,
        billingUnit,
        species,
        note,
        amount: 0,
      };
      current.amount += Number(item.payableAmount ?? item.amount ?? 0);
      details.set(key, current);
    }),
  );
  if (!details.size) return "";
  const linesMarkup = [...details.values()]
    .sort((left, right) =>
      `${left.iacuc}|${left.startDate}|${left.endDate}`.localeCompare(
        `${right.iacuc}|${right.startDate}|${right.endDate}`,
        "zh-CN",
      ),
    )
    .map((item, index) => {
      const unit = item.billingUnit === "animal_day" ? "只" : "笼";
      const note = item.note ? `。${item.note}` : "";
      return noteEntryMarkup(
        index === 0 ? "自定义收费：" : "",
        `${item.iacuc}：${item.startDate || "-"} 至 ${item.endDate || "-"}，每日${numberText(item.quantity)}${unit}${item.species}，${numberText(item.unitPrice)}元/${unit}/日，本月共计${money(item.amount)}元${note}`,
      );
    })
    .join("");
  return linesMarkup;
}

function noteEntryMarkup(title: string, detail: string) {
  return `<div class="note-entry">${title ? `<strong>${escapeHtml(title)}</strong>` : ""}<span class="note-detail">${escapeHtml(detail)}</span></div>`;
}

function rateDescriptor(column: SettlementNoteColumn) {
  const unit = column.billingUnit === "animal_day" ? "只" : "笼";
  if (
    column.speciesLabel === "小鼠" &&
    column.tiered &&
    column.hasTieredCharge &&
    column.overageUnitPrice > 0 &&
    column.overageUnitPrice !== column.unitPrice
  ) {
    return `笼位数≤160，${formatNumber(column.unitPrice)}元/笼/日；笼位数＞160，${formatNumber(column.overageUnitPrice)}元/笼/日`;
  }
  return `${formatNumber(column.unitPrice)}元/${unit}/日`;
}

function formatNumber(value: number) {
  return Number.isInteger(value) ? String(value) : String(Number(value.toFixed(2)));
}
